### PR TITLE
Add Cloud Storage mode of Agent Sidecar.

### PR DESCRIPTION
**Pull request by hlts2**
### Environment:

I added Cloud Storage mode of agent sidecar.
The following libraries have been added for implementation.
~~`go get github.com/google/go-cloud`~~
`go get gocloud.dev`

related project card is this.
https://github.com/vdaas/vald/projects/1#card-39749921

- Go Version: 1.14.3
- Docker Version: 19.03.8
- Kubernetes Version: 1.18.2
- NGT Version: 1.11.5

### Types of changes:

- [ ] Bug fix [type/bug]
- [x] New feature [type/feature]
- [ ] Add tests [type/test]
- [ ] Security related changes [type/security]
- [ ] Add documents [type/documentation]
- [ ] Refactoring [type/refactoring]
- [ ] Update dependencies [type/dependency]
- [ ] Update benchmarks and performances [type/bench]
- [ ] Update CI [type/ci]

### Changes to Core Features:

- [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your core changes, as applicable?
- [ ] Have you successfully ran tests with your changes locally?

### Checklist:

- [x] I have read the [CONTRIBUTING](https://github.com/vdaas/vald/blob/master/CONTRIBUTING.md) document.
- [x] I have checked open [Pull Requests](https://github.com/vdaas/vald/pulls) for the similar feature or fixes?
- [ ] I have added tests and benchmarks to cover my changes.
- [ ] I have ensured all new and existing tests passed.
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: internal/db/storage/blob/cloudstorage/cloudstorage.go [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+	r, err := reader.New(
+		reader.WithKey(key),
+		reader.WithBucket(c.bucket),
+		reader.WithReaderOptions(c.readerOpts),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	return r, r.Open(ctx)
```
Comment (rinx) [inline internal/db/storage/blob/cloudstorage/cloudstorage.go:63]: badge
i think there's no need to create `reader` package.
```suggestion
	return c.bucket.NewReader(ctx, key, c.readerOpts)
```

This is just my opinion. If you'd like to do that, it is okay.

File: internal/db/storage/blob/cloudstorage/cloudstorage.go [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+	w, err := writer.New(
+		writer.WithKey(key),
+		writer.WithBucket(c.bucket),
+		writer.WithWriterOptions(c.writerOpts),
+	)
+	if err != nil {
+		return nil, err
+	}
+	return w, w.Open(ctx)
```
Comment (rinx) [inline internal/db/storage/blob/cloudstorage/cloudstorage.go:79]: badge
```suggestion
	return c.bucket.NewWriter(ctx, key, c.writerOpts)
```

File: internal/db/storage/blob/cloudstorage/cloudstorage.go [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+	for _, opt := range append(defaultOpts, opts...) {
+		if err := opt(c); err != nil {
+			return nil, err
+		}
+	}
+	return c, nil
+}
+
+func (c *client) Open(ctx context.Context) (err error) {
+	c.bucket, err = blob.OpenBucket(ctx, c.urlstr)
```
Comment (rinx) [inline internal/db/storage/blob/cloudstorage/cloudstorage.go:36]: badge
Please use the `OpenBucketURL` function in gcsblob package instead.
https://godoc.org/gocloud.dev/blob/gcsblob#URLOpener.OpenBucketURL

File: internal/db/storage/blob/cloudstorage/cloudstorage.go (hunk 1 above)
Comment (hlts2) [reply internal/db/storage/blob/cloudstorage/cloudstorage.go:63]: I think your opinion is good. 

However, it is difficult to test with the current structure, so can I make c.buckt a private interface?
the following is `example`.
```go
type bucket interface {
     NewReader(ctx context.Context, key string, opts *ReaderOptions) (*Reader, error)
     NewWriter(ctx context.Context, key string, opts *WriterOptions) (*Writer, err error)
     Close()
}

```

File: internal/db/storage/blob/cloudstorage/cloudstorage.go (hunk 2 above)
Comment (hlts2) [reply internal/db/storage/blob/cloudstorage/cloudstorage.go:79]: same

https://github.com/vdaas/vald/pull/519#discussion_r445985137

File: internal/db/storage/blob/cloudstorage/cloudstorage.go (hunk 3 above)
Comment (hlts2) [reply internal/db/storage/blob/cloudstorage/cloudstorage.go:36]: Thank you for your review.
I will fix it soon

File: internal/db/storage/blob/cloudstorage/cloudstorage.go (hunk 1 above)
Comment (rinx) [reply internal/db/storage/blob/cloudstorage/cloudstorage.go:63]: It is okay to add a new interface, but i think it is much easier to use the memblob package for test use.
https://godoc.org/gocloud.dev/blob/memblob

Comment (hlts2) [reply internal/db/storage/blob/cloudstorage/cloudstorage.go:63]: Thank you for your suggestion.
I did not know this pacakge. 
Thank you and I use it.

File: charts/vald/values.yaml [hunk 4]
```diff
@@ -1445,6 +1445,43 @@ agent:
...
+            # @schema {"name": "agent.sidecar.config.blob_storage.cloud_storage.client.private_key", "type": "string"}
+            # agent.sidecar.config.blob_storage.cloud_storage.client.private_key -- private key
+            private_key: ""
+            # @schema {"name": "agent.sidecar.config.blob_storage.cloud_storage.client.credentials_file_path", "type": "string"}
+            # agent.sidecar.config.blob_storage.cloud_storage.client.credentials_file_path -- credentials file path
+            credentials_file_path: ""
+            # @schema {"name": "agent.sidecar.config.blob_storage.cloud_storage.client.credentials_json", "type": "string"}
+            # agent.sidecar.config.blob_storage.cloud_storage.client.credentials_json -- credentials json
+            credentials_json: ""
+          # @schema {"name": "agent.sidecar.config.blob_storage.cloud_storage.write_buffer_size", "type": "string"}
```
Comment (rinx) [inline charts/vald/values.yaml:1467]:
```suggestion
          # @schema {"name": "agent.sidecar.config.blob_storage.cloud_storage.write_buffer_size", "type": "integer"}
```

Comment (hlts2) [reply charts/vald/values.yaml:1467]: sorry :bow:

**Comment by hlts2** [comment]
I couldn't rebase, so I chose merge commit. I will be careful from the next time.

**Comment by hlts2** [comment]
@kpango @rinx 
if you have a time, could you please review this PR?:bow: (~4/19)
It is not an urgent matter.  :bow:
thank you always.

File: pkg/agent/sidecar/usecase/initcontainer/initcontainer.go [hunk 5]
```diff
@@ -252,5 +268,8 @@ func (r *run) Stop(ctx context.Context) error {
 }
 
 func (r *run) PostStop(ctx context.Context) error {
+	if r.rs != nil {
+		r.rs.PostStop(ctx)
```
Comment (rinx) [inline pkg/agent/sidecar/usecase/initcontainer/initcontainer.go:272]: should be in PreStop.

File: pkg/agent/sidecar/service/restorer/restorer.go [hunk 6]
```diff
@@ -111,6 +112,13 @@ func (r *restorer) Start(ctx context.Context) (<-chan error, error) {
 	return ech, nil
 }
 
+func (r *restorer) PostStop(ctx context.Context) error {
+	if r.storage != nil {
+		return r.storage.Stop(ctx)
+	}
```
[the remark this answers is not in the document]
Comment (hlts2) [reply pkg/agent/sidecar/service/restorer/restorer.go:118]: Thank you.:bow:
Got it!
I fixed: https://github.com/vdaas/vald/pull/519/commits/a7f2c670c4bec8c20342972b37d4757089c3fa52

File: pkg/agent/sidecar/usecase/initcontainer/initcontainer.go (hunk 5 above)
Comment (hlts2) [reply pkg/agent/sidecar/usecase/initcontainer/initcontainer.go]: Thank you.:bow:
Got it!
I fixed: https://github.com/vdaas/vald/pull/519/commits/a7f2c670c4bec8c20342972b37d4757089c3fa52

**Comment by hlts2** [comment]
@rinx 
thank you for review.
I fixed everything.
if you have a time, could you please re-review?

File: pkg/agent/sidecar/usecase/initcontainer/initcontainer.go [hunk 7]
```diff
@@ -241,6 +257,9 @@ func (r *run) Start(ctx context.Context) (<-chan error, error) {
 }
 
 func (r *run) PreStop(ctx context.Context) error {
+	if r.rs != nil {
+		r.rs.PreStop(ctx)
```
[the remark this answers is not in the document]
Comment (hlts2) [reply pkg/agent/sidecar/usecase/initcontainer/initcontainer.go]: @rinx 
sorry.
I got a warning, so I fixed it. 
https://github.com/vdaas/vald/pull/519/commits/7621d65c27a2a0bdbe7821b4b1ffc987f3ffb23e

**Review by rinx — APPROVED** [review]

**Comment by hlts2** [comment]
@kpango 
I am sorry for the reminder:bow:
I know that you are busy. 
if you have a time, could you please review this pr?

https://github.com/vdaas/vald/pull/519#issuecomment-817417120

**Comment by kpango** [comment]
can you please add author-review tag?

**Comment by hlts2** [comment]
@kpango 
sorry. I forgot to add label...
I added it:bow:

**Comment by kpango** [comment]
Reviewing this PR is the lowest priority of my tasks now, and it takes a lot of time, but if you can add the author-reivew tag, I can look at that tag and review it when I have time.

**Comment by hlts2** [comment]
@kpango I got it. thank you.

**Review by kpango — APPROVED** [review]
LGTM

**Pull request merged**